tools/walletextension/wallet_extension.go

```diff
@@ -153,105 +131,111 @@ func (we *WalletExtension) handleHTTPEthJSON(resp http.ResponseWriter, req *http
-		return
-	}
+	respMap := make(map[string]interface{})
+	respMap[resJSONKeyID] = reqID
+	respMap[resJSONKeyRPCVer] = jsonrpc.Version
 
-	// We decrypt the result field if it's encrypted.
-	maybeDecryptedRespJSONMap, err := we.decryptResponseIfNeeded(method, respJSONMap)
-	if err != nil {
-		logAndSendErr(resp, fmt.Sprintf("could not decrypt response: %s", err))
-		return
+	// respStr could be a plain string value or could be a json encoded object
+	respMap[RespJSONKeyResult] = respStr
+
+	// we try to unmarshal respStr into json and set the object in the structure if it's valid json
+	var respObj json.RawMessage
+	err = json.Unmarshal([]byte(respStr), &respObj)
+	if err == nil {
+		// unmarshal successful, set the object in response
+		respMap[RespJSONKeyResult] = respObj
 	}
 
-	// We marshal the response to present to the client.
-	clientResponse, err := json.Marshal(maybeDecryptedRespJSONMap)
+	rpcResponse, err := json.Marshal(respMap)
 	if err != nil {
-		logAndSendErr(resp, fmt.Sprintf("could not marshal JSON response to present to the client: %s", err))
-		return
+		logAndSendErr(resp, fmt.Sprintf("failed to remarshal RPC response to return to caller: %s", err))
 	}
-	fmt.Printf("Received %s response from Obscuro node: %s\n", method, strings.TrimSpace(string(clientResponse)))
 
 	// We write the response to the client.
-	_, err = resp.Write(clientResponse)
+	_, err = resp.Write(rpcResponse)
 	if err != nil {
 		logAndSendErr(resp, fmt.Sprintf("could not write JSON-RPC response: %s", err))
 		return
 	}
 }
 
-// If an `eth_call` request doesn't have a `from` field, we won't be able to encrypt the response. In that case, we use
-// the viewing key address as the `from` field to allow encryption and decryption.
-func (we *WalletExtension) ensureCallsHaveFromField(method interface{}, reqJSONMap map[string]interface{}) (map[string]interface{}, error) {
-	if method != ReqJSONMethodCall {
-		// We only modify `eth_call` requests.
-		return reqJSONMap, nil
+// extract.AndPrepareParams extracts the params from the request to pass into rpc client.
+//	May also modify params etc. as needed for specific methods
+func (we *WalletExtension) extractAndPrepareParams(method string, reqJSONMap map[string]json.RawMessage) ([]interface{}, error) {
+	// we extract the params into a JSON list
+	var params []interface{}
+	err := json.Unmarshal(reqJSONMap[reqJSONKeyParams], &params)
+	if err != nil {
+		return nil, fmt.Errorf("could not unmarshal params list from JSON-RPC request body: %w", err)
 	}
 
-	params, ok := reqJSONMap[reqJSONKeyParams].([]interface{})
-	if !ok {
-		return nil, fmt.Errorf("params for %s request were malformed", method)
+	// If an `eth_call` request doesn't have a `from` field, we won't be able to encrypt the response. In that case, we use
+	// the viewing key address as the `from` field to allow encryption and decryption.
+	if method == ReqJSONMethodCall {
```

Comment:
@BedrockSquirrel I think for ease of reading, this check should be inside the method call to `ensureParamsHaveFromField`.